### DISCUSSION TITLE
Incorrect showing of method/function parameters and 'docstring' information

### DISCUSSION THREAD

**Discussion by ma-galan** [microsoft/pylance-release#2711] [Enhancement]
## Environment data

-   Language Server version: Pylance 2020.7.3
-   OS and version: Windows 7 64bits
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.8.5 64-bit (conda)

## Expected behaviour
for example, in Jupyter Lab:
![Correct01](https://user-images.githubusercontent.com/53902830/88386220-37a94600-cdb0-11ea-8448-b874badcbf56.jpg)

![Correct02](https://user-images.githubusercontent.com/53902830/88386223-3a0ba000-cdb0-11ea-876d-bbcb07513db3.jpg)

## Actual behaviour

Not showing the correct docstring and parameters information. I don't know if this is a problem related to Python extension or Pylance.

![Docs01](https://user-images.githubusercontent.com/53902830/88385851-7a1e5300-cdaf-11ea-8e95-8365e44f8e3a.jpg)

![Docs02](https://user-images.githubusercontent.com/53902830/88385858-7d194380-cdaf-11ea-8b7a-00fed80a6fbe.jpg)

![Docs03](https://user-images.githubusercontent.com/53902830/88385860-7f7b9d80-cdaf-11ea-8e1b-2351a40fa09a.jpg)

![Docs04](https://user-images.githubusercontent.com/53902830/88385866-81ddf780-cdaf-11ea-99ba-e73f64fb5ce7.jpg)

## Logs

```
XXX
```

## Code Snippet / Additional information

```python
XXX
```

**Comment by jakebailey** [comment]
Pylance is a static analyzer, and can't get that information at runtime like Jupyter does (which has already executed the code and can access the real info).

In that first case, am I correct in saying that it's pandas? Our pandas stubs include the docstrings (as the "real" code produces them extremely dynamically), so it's possible some are missing.

As for `open`, that's #49.

**Comment by ma-galan** [comment]
Thanks for the quick reply.
You are right, the first case is Pandas. I checked different pandas methods and none of them is shown properly (parameters are shown in an odd way and docstring is not shown):

![Docs05](https://user-images.githubusercontent.com/53902830/88432209-0f473900-cdfc-11ea-9f24-af1d73dd0694.jpg)

![Docs06](https://user-images.githubusercontent.com/53902830/88432216-140bed00-cdfc-11ea-8d22-73380ce4dedf.jpg)

![Docs07](https://user-images.githubusercontent.com/53902830/88432233-1b32fb00-cdfc-11ea-8d54-1655ed89d211.jpg)

**Comment by jakebailey** [comment]
Thanks. In all of those cases, you're in the "signature help" which is missing the docstrings in our stubs (and we apparently can't find it in the real implementation).

@huguesv will be able to look at this later.

Out of curiosity, what are you finding "odd" about the parameters?

**Comment by ma-galan** [comment]
Sorry, I meant that it shows all the parameter options without a 'nice' format, which is what is nicely presented in a docstring. For me the best way to show parameters and docstring is the way that Jupyter Lab represents it (as per the screenshots shown in my first post).

**Comment by huguesv** [comment]
We'll need to reorganize our pandas stubs slightly to have the `DataFrame` class (among others) be in a relative location in the stub that matches the location in sources.

Right now, in stub we currently have `DataFrame` in `pandas.core`, but in sources it's in `pandas.core.frame`, so Pylance is not finding the class implementation in `pandas.core`.

Doing this should fix the tooltip, as well as go to definition.

**Comment by gramster** [comment]
So @huguesv in the next version the stubs will reflect the correct file layout but will still not show these docstrings, so they are likely cases where the docs are synthesized as part of pandas build.

**Comment by cpbotha** [comment]
I am running into this issue with Django as well.

PyLance is showing the docstring-less function signature from django-stubs (which it comes bundled with) instead of the docstring in the actual Django source code.

What I gather from https://github.com/microsoft/pylance-release/issues/170#issuecomment-663719941 above is that PyLance *should* use type stubs for type info and source code for docstrings. However, with Django (for example) this is clearly not the case.

For `django.db.models.ForeignKey`, as one example, PyLance in VSCode shows only the signature on hover:

![image](https://user-images.githubusercontent.com/937871/92112119-99b48c80-eded-11ea-90e8-4838e1988cab.png)

Although the source clearly has a docstring, and VSCode even offers to take me there with `Go to definition`:

![image](https://user-images.githubusercontent.com/937871/92112220-c1a3f000-eded-11ea-80ab-fa4f041dfa09.png)

Ideally hover would show the type information but also any docstring.

**Comment by jakebailey** [comment]
Sorry to give you the rigmarole, but I think the django thing is another thing entirely, either that hovers are missing docstrings or that there's a specific bug with django given it's finding the def.

This issue is about pandas docstrings being generated at runtime where the static analysis can't possibly put together what the docstrings going to be (e.g. they do something like `read_csv.__doc__ = "...".format(...)`.

To save you from writing that all over again, I made #316. 🙂

**Comment by memeplex** [comment]
I'm not sure whether this is the same issue as the one described above but for builtin modules I also find the behavior of PyLance very inconvenient (vs Jedi). For example, this is the documentation popup for `ArgumentParser`:

![image](https://user-images.githubusercontent.com/2845433/94068087-1de2aa00-fdc5-11ea-9fdf-1e484d3bccc6.png)

It's badly formatted but that's not my point. As soon as I add parentheses it becomes:

![image](https://user-images.githubusercontent.com/2845433/94068203-436fb380-fdc5-11ea-99b9-614d185d336b.png)

Notice this is not the popup that is shown while I'm typing arguments but the documentation popup shown on hover (the mouse pointer being above `ArgumentParser`), that is in the same situation as before. I find it next to useless TBH.

**Comment by memeplex** [comment]
Oh, I guess mine is this one https://github.com/microsoft/pylance-release/issues/316#issuecomment-697056063 instead.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
